pkg/policy/rules.go

```diff
@@ -252,3 +253,29 @@ egressLoop:
 
 	return egressDecision
 }
+
+// analyzeWhetherRulesSelectEndpoints iterates over a given list of rules to
+// update the cache within the rule which determines whether or not the given
+// identity is selected by that rule. If a rule in the list does select said
+// identity, it is added to epIDSet. Signals to the given WaitGroup that
+// all rules have been parsed in relation to said identity.
+func (rules ruleSlice) analyzeWhetherRulesSelectEndpoint(identityConsumer IdentityConsumer, epIDSet *IDSet, wg *sync.WaitGroup) {
+	if identityConsumer == nil {
+		log.Warning("identity consumer is nil; returning")
+		wg.Done()
+		return
+	}
+	id := identityConsumer.GetID16()
+	securityIdentity := identityConsumer.GetSecurityIdentity()
+
+	for _, r := range rules {
+		if ruleMatches := r.matches(id, securityIdentity); ruleMatches {
+			epIDSet.Mutex.Lock()
+			epIDSet.IDs[id] = struct{}{}
+			epIDSet.Mutex.Unlock()
```

Comment:
This locking seems off. The epIDSet is shared across many goroutines, OK, so there's a mutex. But then every time there's a rule match, an endpoint will block out updating the IDSet from any other goroutine. And after the first rule match, this IDSet update is a no-op. With a ruleset that matches many endpoints, the goroutines will basically be serialized.

Reply:
Was wondering about a better way to signal this back up, but honestly it's probably OK as-is. It'd be nice to follow up afterwards though, and take a second look at how this function is used for the different cases I highlighted in the comment on this function's signature (cache update / determine if EP needs regeneration) and check that the work being done here is reasonable.

Reply:
I definitely could break if the `ID` is already in the set. I should actually check whether the ID is in `epIDSet` before iterating over the rules, as it is updated across goroutines.

Reply:
Right, so that's why I'm suspecting you need two separate paths here - one that deliberately attempts to update the caches everywhere, and one that does the "endpoint regeneration" determination, and short-cuts when possible. (Or maybe you just need the latter, and rely on lazy evaluation to set up the cache).

Reply:
@joestringer see 
https://github.com/cilium/cilium/pull/7465 - I think this does what you desire.